### PR TITLE
CNF-14084: controller: update MachineConfig reconciliation

### PR DESCRIPTION
**Pull request by Tal-or**
Starting from 4.18 RTE pods can use a built-in SELinux policy instead
of custom one.

This means that MachineConfig deployment is not mandatory anymore.
However, it is still an option to deploy a custom policy using MachineConfig.

This commit updates the MahcineConfig reconciliation logic.

By default when upgrade is done from 4.1X -> 4.18 the controller
tries to remove the redundant MahcineConfig, unless user state
explicitly that custom policy is needed.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

### CODE REVIEW

**Comment by ffromani** [comment]
> Starting from 4.18 RTE pods can use a built-in SELinux policy instead of custom one.
> 
> This means that MachineConfig deployment is not mandatory anymore. However, it is still an option to deploy a custom policy using MachineConfig.
> 
> This commit updates the MahcineConfig reconciliation logic.
> 
> By default when upgrade is done from 4.1X -> 4.18 the controller tries to remove the redundant MahcineConfig, unless user state explicitly that custom policy is needed.

this could be the bit that pushes us to v2.
New deployments should use the new builtin policy
Old deployments should keep using the old one till they upgrade explicitely

But then we have conflicting defaults :\

**Comment by Tal-or** [comment]
> > Starting from 4.18 RTE pods can use a built-in SELinux policy instead of custom one.
> > This means that MachineConfig deployment is not mandatory anymore. However, it is still an option to deploy a custom policy using MachineConfig.
> > This commit updates the MahcineConfig reconciliation logic.
> > By default when upgrade is done from 4.1X -> 4.18 the controller tries to remove the redundant MahcineConfig, unless user state explicitly that custom policy is needed.
> 
> this could be the bit that pushes us to v2. New deployments should use the new builtin policy Old deployments should keep using the old one till they upgrade explicitely
> 
> But then we have conflicting defaults :\

Let's discuss it more deeply offline

File: internal/api/features/types.go [hunk 1]
```diff
@@ -23,7 +23,8 @@ import (
 )
 
 const (
-	Version = "v4.18.0"
+	Version                            = "v4.18.0"
+	ForceCustomSELinuxPolicyAnnotation = "force-custom-selinux-policy"
```
Comment (Tal-or) [inline internal/api/features/types.go:27]: The API/name is open for discussion.

File: api/numaresourcesoperator/v1/numaresourcesoperator_types.go [hunk 2]
```diff
@@ -189,3 +190,8 @@ func (ngc *NodeGroupConfig) ToString() string {
 	}
 	return ""
 }
+
+func (nro *NUMAResourcesOperator) IsCustomPolicyEnabled() bool {
+	_, ok := nro.Annotations[features.ForceCustomSELinuxPolicyAnnotation]
+	return ok
+}
```
Comment (ffromani) [inline api/numaresourcesoperator/v1/numaresourcesoperator_types.go:197]: this should be a free function in a different package, perhaps `internal/api/annotations` again

Comment (ffromani) [reply api/numaresourcesoperator/v1/numaresourcesoperator_types.go:197]: The rationale is that all the magic annotation constants should not be exposed in the public packages. Users will see only the magic strings, but they won't be part of the API contracts now (preferibily, never)

File: api/numaresourcesoperator/v1/numaresourcesoperator_types.go [hunk 3]
```diff
@@ -22,6 +22,7 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	"github.com/openshift-kni/numaresources-operator/internal/api/features"
```
Comment (ffromani) [inline api/numaresourcesoperator/v1/numaresourcesoperator_types.go:25]: let's avoid this, will make the package non importable: https://docs.google.com/document/d/1e8kOo3r51b2BWtTs_1uADIA5djfXhPT36s6eHVRIvaU/edit

File: controllers/numaresourcesoperator_controller.go [hunk 4]
```diff
@@ -318,41 +320,54 @@ func (r *NUMAResourcesOperatorReconciler) syncNodeResourceTopologyAPI(ctx contex
...
-		if err2 := controllerutil.SetControllerReference(instance, objState.Desired, r.Scheme); err2 != nil {
-			err = errors.Wrapf(err2, "failed to set controller reference to %s %s", objState.Desired.GetNamespace(), objState.Desired.GetName())
-			break
+	// Since 4.18 we're using a built-in SELinux policy,
+	// so the MachineConfig which applies the custom policy is no longer necessary.
+	// In case of operator upgrade from 4.1X → 4.18, it's necessary to remove the old MachineConfig,
+	// unless an emergency annotation is provided which forces the operator to use custom policy
+	if !instance.IsCustomPolicyEnabled() {
+		for _, objState := range objStates {
+			if !objState.IsNotFoundError() {
```
Comment (ffromani) [inline controllers/numaresourcesoperator_controller.go:340]: this looks fragile. We want to target only the `MachineConfig`. So it's probably better to add new functionalities to `pkg/objectstate/rte`

File: controllers/numaresourcesoperator_controller.go [hunk 5]
```diff
@@ -318,41 +320,54 @@ func (r *NUMAResourcesOperatorReconciler) syncNodeResourceTopologyAPI(ctx contex
...
+			}
 
-		_, updated, err2 := apply.ApplyObject(ctx, r.Client, objState)
-		if err2 != nil {
-			err = errors.Wrapf(err2, "could not apply (%s) %s/%s", objState.Desired.GetObjectKind().GroupVersionKind(), objState.Desired.GetNamespace(), objState.Desired.GetName())
-			break
-		}
-		if !updated {
-			continue
+			_, _, err2 := apply.ApplyObject(ctx, r.Client, objState)
```
Comment (ffromani) [inline controllers/numaresourcesoperator_controller.go:361]: let's see if we can add `apply.DeleteObject`. I'll think about it

File: controllers/numaresourcesoperator_controller.go [hunk 6]
```diff
@@ -318,41 +320,54 @@ func (r *NUMAResourcesOperatorReconciler) syncNodeResourceTopologyAPI(ctx contex
...
+			if err2 != nil {
+				err = errors.Wrapf(err2, "could not apply (%s) %s/%s", objState.Desired.GetObjectKind().GroupVersionKind(), objState.Desired.GetNamespace(), objState.Desired.GetName())
+				break
+			}
 		}
-		updatedCount++
+		mcpUpdatedFunc = IsMachineConfigPoolUpdated
```
Comment (ffromani) [inline controllers/numaresourcesoperator_controller.go:367]: nice idea

File: pkg/objectstate/rte/rte.go [hunk 7]
```diff
@@ -91,19 +91,17 @@ func (em *ExistingManifests) MachineConfigsState(mf rtemanifests.Manifests) []ob
...
-			ret = append(ret,
-				objectstate.ObjectState{
-					Existing: existingMachineConfig.machineConfig,
-					Error:    existingMachineConfig.machineConfigError,
-					Desired:  desiredMachineConfig,
-					Compare:  compare.Object,
-					Merge:    merge.ObjectForUpdate,
-				},
-			)
+			ret = append(ret, os)
```
Comment (ffromani) [inline pkg/objectstate/rte/rte.go:102]: why this change?

File: controllers/numaresourcesoperator_controller.go [hunk 8]
```diff
@@ -771,6 +773,15 @@ func daemonsetUpdater(mcpName string, gdm *rtestate.GeneratedDesiredManifest) er
 		klog.V(5).InfoS("DaemonSet update: cannot update config", "mcp", mcpName, "daemonset", gdm.DaemonSet.Name, "error", err)
 		return err
 	}
+	if gdm.ClusterPlatform != platform.Kubernetes {
+		if instance.IsCustomPolicyEnabled() && gdm.ClusterPlatform == platform.OpenShift {
```
Comment (ffromani) [inline controllers/numaresourcesoperator_controller.go:778]: I'd rather move the flag into the `GeneratedDesiredManifest` struct

File: pkg/objectstate/rte/rte.go [hunk 9]
```diff
@@ -130,13 +130,13 @@ type GeneratedDesiredManifest struct {
...
 
-type GenerateDesiredManifestUpdater func(mcpName string, gdm *GeneratedDesiredManifest) error
+type GenerateDesiredManifestUpdater func(mcpName string, gdm *GeneratedDesiredManifest, instance *nropv1.NUMAResourcesOperator) error
 
 func SkipManifestUpdate(gdm *GeneratedDesiredManifest) error {
 	return nil
 }
 
-func (em *ExistingManifests) State(mf rtemanifests.Manifests, updater GenerateDesiredManifestUpdater) []objectstate.ObjectState {
+func (em *ExistingManifests) State(mf rtemanifests.Manifests, instance *nropv1.NUMAResourcesOperator, updater GenerateDesiredManifestUpdater) []objectstate.ObjectState {
```
Comment (ffromani) [inline pkg/objectstate/rte/rte.go:139]: `FromClient` already gets a `instance` parameter, I think we can safely compute the flag there and carry it in the `ExistingManifests` struct

File: test/e2e/install/install_test.go [hunk 10]
```diff
@@ -288,7 +288,7 @@ var _ = Describe("[Install] durability", func() {
 			By("checking there are no leftovers")
 			// by taking the ns from the ds we're avoiding the need to figure out in advanced
 			// at which ns we should look for the resources
-			mf, err := rte.GetManifests(configuration.Plat, configuration.PlatVersion, ds.Namespace, true)
+			mf, err := rte.GetManifests(configuration.Plat, configuration.PlatVersion, ds.Namespace, true, true)
```
Comment (ffromani) [inline test/e2e/install/install_test.go:294]: time to cleanup this signature if time allows

File: test/utils/deploy/deploy.go [hunk 11]
```diff
@@ -37,6 +37,7 @@ import (
 	"github.com/openshift-kni/numaresources-operator/controllers"
 	nropmcp "github.com/openshift-kni/numaresources-operator/internal/machineconfigpools"
 	"github.com/openshift-kni/numaresources-operator/pkg/status"
+	"github.com/openshift-kni/numaresources-operator/test/e2e/serial/tests"
```
Comment (ffromani) [inline test/utils/deploy/deploy.go:40]: please no: code in `test/e2e/serial/tests` may depend on these utilities but not the other way around

File: controllers/numaresourcesoperator_controller.go [hunk 12]
```diff
@@ -478,6 +478,7 @@ func (r *NUMAResourcesOperatorReconciler) syncNUMAResourcesOperatorResources(ctx
 		}
 		rteupdate.DaemonSetHashAnnotation(r.RTEManifests.DaemonSet, cmHash)
 	}
+	rteupdate.SecurityContextConstraintSetSELinuxOption(r.RTEManifests.SecurityContextConstraint, instance.IsCustomPolicyEnabled())
```
Comment (ffromani) [inline controllers/numaresourcesoperator_controller.go:481]: maybe just `rteupdate.SecurityContextConstraint()`

**Review by ffromani — COMMENTED** [review]
most comments are about code organization. The overall direction is reasonnable, but we would need some work still

File: controllers/numaresourcesoperator_controller.go (hunk 4 above)
Comment (ffromani) [reply controllers/numaresourcesoperator_controller.go:340]: I'm also thinking about `deleteUnusedMachineConfigs`. We can perhaps reuse that function - or even better get rid of it.

Comment (ffromani) [reply controllers/numaresourcesoperator_controller.go:340]: ...or not? https://github.com/openshift-kni/numaresources-operator/pull/1029

File: api/numaresourcesoperator/v1/numaresourcesoperator_types.go (hunk 3 above)
Comment (Tal-or) [reply api/numaresourcesoperator/v1/numaresourcesoperator_types.go:25]: Here: https://github.com/openshift-kni/numaresources-operator/pull/1025#discussion_r1784578995 you mentioned you want it non-importable, I'm a bit confused :)

Comment (ffromani) [reply api/numaresourcesoperator/v1/numaresourcesoperator_types.go:25]: I could have been clearer. The magic annotations should not be part of the API packages, nor of its helpers.
Thus, let's move the helpers away from the api packages: rather than import the internal package in the api package, let's move the helper in the internal package.

File: controllers/numaresourcesoperator_controller.go (hunk 4 above)
Comment (Tal-or) [reply controllers/numaresourcesoperator_controller.go:340]: What you did in #1029 is true, but we still need to remove the machine config explicitly when moving to built-in policy, because the CR remain in the cluster.
And as long as it persist on the cluster the machine-config won't go anywhere

Comment (ffromani) [reply controllers/numaresourcesoperator_controller.go:340]: OK, let's try to remove them in the reconciliation loop rather than explicitely in the `deleteXXX` functions.

Comment (Tal-or) [reply controllers/numaresourcesoperator_controller.go:340]: > this looks fragile. We want to target only the MachineConfig. So it's probably better to add new functionalities to pkg/objectstate/rte

We do. you have the `objStates := existing.MachineConfigsState(r.RTEManifests)` that filters the MachineConfigs out of all the objects for you.

> OK, let's try to remove them in the reconciliation loop rather than explicitely in the deleteXXX functions.

This code is part of the reconciliation loop
it's called in `syncMachineConfigs`

File: controllers/numaresourcesoperator_controller.go (hunk 6 above)
Comment (Tal-or) [reply controllers/numaresourcesoperator_controller.go:367]: Thanks!

File: controllers/numaresourcesoperator_controller.go (hunk 4 above)
Comment (ffromani) [reply controllers/numaresourcesoperator_controller.go:340]: yes, I commented before to starting the effort in #1029 . It seems this of yours is the best approach on the table, and the removal of the `deleteXXX` functions should be tried (and preferably done!) on the side

File: internal/api/features/types.go (hunk 1 above)
Comment (Tal-or) [reply internal/api/features/types.go]: Done.

File: api/numaresourcesoperator/v1/numaresourcesoperator_types.go (hunk 2 above)
Comment (Tal-or) [reply api/numaresourcesoperator/v1/numaresourcesoperator_types.go]: Done. Thanks!

File: pkg/objectstate/rte/rte.go (hunk 7 above)
Comment (Tal-or) [reply pkg/objectstate/rte/rte.go]: residues of an old implementation effort. Removed the changes

File: controllers/numaresourcesoperator_controller.go (hunk 8 above)
Comment (Tal-or) [reply controllers/numaresourcesoperator_controller.go]: Done.

File: pkg/objectstate/rte/rte.go (hunk 9 above)
Comment (Tal-or) [reply pkg/objectstate/rte/rte.go]: Done.

File: test/e2e/install/install_test.go (hunk 10 above)
Comment (Tal-or) [reply test/e2e/install/install_test.go:294]: Will do, but I wouldn't block this PR for that.

File: test/utils/deploy/deploy.go (hunk 11 above)
Comment (Tal-or) [reply test/utils/deploy/deploy.go]: Done. I moved the MCP wait logic to `test/utils/deploy/deploy.go` and 
now `github.com/openshift-kni/numaresources-operator/test/e2e/serial/tests` consumes it

File: test/e2e/install/install_test.go (hunk 10 above)
Comment (ffromani) [reply test/e2e/install/install_test.go:294]: of course no, this must to be a separate independent effort

**Review by ffromani — COMMENTED** [review]
/approve
/lgtm

I'm not completely happy with how the code would look like, but I prefer to merge the functionality and improve in followup PRs.

@Tal-or please unhold if you're happy with this work and if you think the test coverage is sufficient

**Comment by Tal-or** [comment]
> /approve /lgtm
> 
> I'm not completely happy with how the code would look like, but I prefer to merge the functionality and improve in followup PRs.
> 
> @Tal-or please unhold if you're happy with this work and if you think the test coverage is sufficient

We can wait and have another iteration if you're not happy with code flow. I don't mind at all

**Comment by ffromani** [comment]
> > /approve /lgtm
> > I'm not completely happy with how the code would look like, but I prefer to merge the functionality and improve in followup PRs.
> > @Tal-or please unhold if you're happy with this work and if you think the test coverage is sufficient
> 
> We can wait and have another iteration if you're not happy with code flow. I don't mind at all

for starters, I'd love to push the Delete logic in the `apply` package, for consistency. Something like https://github.com/openshift-kni/numaresources-operator/pull/1038
